From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pekka Vuorela <pekka.vuorela@jolla.com>
Date: Wed, 8 Nov 2023 13:39:37 +0200
Subject: [PATCH] Introduce QMailAccount::HasPersistentConnection status flag

Use QMailAccount::HasPersistentConnection status flag for IMAP idle instead of
updating last sync time every minute, this reduces accounts db writes(they
trigger notifications to other processes) and also make use of a linger
keepalive(less state transitions).
---
 src/libraries/qmfclient/qmailstore_p.cpp | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/libraries/qmfclient/qmailstore_p.cpp b/src/libraries/qmfclient/qmailstore_p.cpp
index cfc1b04f..e8194303 100644
--- a/src/libraries/qmfclient/qmailstore_p.cpp
+++ b/src/libraries/qmfclient/qmailstore_p.cpp
@@ -2675,6 +2675,10 @@ bool SSOAccountSatisfyTheProperty(Accounts::Account* ssoAccount, const QMailAcco
             status &= (~QMailAccount::AppendSignature);
             status |= appendSignature?(QMailAccount::AppendSignature):0;
 
+            bool hasPersistentConnection = ssoAccount->valueAsBool(QLatin1String("hasPersistentConnection"), false);
+            status &= (~QMailAccount::HasPersistentConnection);
+            status |= hasPersistentConnection?(QMailAccount::HasPersistentConnection):0;
+
             return SSOAccountCompareProperty<quint64>(ssoAccount,
                                                       status,
                                                       argument.op, argument.valueList);
@@ -3719,11 +3723,13 @@ QMailAccount QMailStorePrivate::extractAccount(const QSharedPointer<Accounts::Ac
     const bool& isDefault = ssoAccount->valueAsBool(QLatin1String("email/default"));
     const bool& canTransmit = ssoAccount->valueAsBool(QLatin1String("canTransmit"), true);
     const bool& appendSignature = ssoAccount->valueAsBool(QLatin1String("signatureEnabled"), true);
+    const bool& hasPersistentConnection = ssoAccount->valueAsBool(QLatin1String("hasPersistentConnection"), false);
 
     result.setStatus(QMailAccount::Enabled, enabled);
     result.setStatus(QMailAccount::PreferredSender, isDefault);
     result.setStatus(QMailAccount::CanTransmit, canTransmit);
     result.setStatus(QMailAccount::AppendSignature, appendSignature);
+    result.setStatus(QMailAccount::HasPersistentConnection, hasPersistentConnection);
 
     result.setSignature(ssoAccount->valueAsString(QLatin1String("signature")));
     result.setFromAddress(ssoAccount->contains(QLatin1String("fullName"))
@@ -6227,6 +6233,8 @@ QMailStorePrivate::AttemptResult QMailStorePrivate::attemptAddAccount(QMailAccou
     ssoAccount->setValue(QLatin1String("status"), account->status());
     const bool appendSignature = (account->status() & QMailAccount::AppendSignature);
     ssoAccount->setValue(QLatin1String("signatureEnabled"), appendSignature);
+    const bool hasPersistentConnection = (account->status() & QMailAccount::HasPersistentConnection);
+    ssoAccount->setValue(QLatin1String("hasPersistentConnection"), hasPersistentConnection);
     ssoAccount->setValue(QLatin1String("signature"), account->signature());
     ssoAccount->setValue(QLatin1String("emailaddress"), account->fromAddress().address());
     ssoAccount->setValue(QLatin1String("fullName"), account->fromAddress().name());
@@ -7028,6 +7036,8 @@ QMailStorePrivate::AttemptResult QMailStorePrivate::attemptUpdateAccount(QMailAc
         ssoAccount->setValue(QLatin1String("status"), account->status());
         bool signatureEnabled = account->status() & QMailAccount::AppendSignature;
         ssoAccount->setValue(QLatin1String("signatureEnabled"), signatureEnabled);
+        bool hasPersistentConnection = account->status() & QMailAccount::HasPersistentConnection;
+        ssoAccount->setValue(QLatin1String("hasPersistentConnection"), hasPersistentConnection);
         ssoAccount->setValue(QLatin1String("signature"), account->signature());
         ssoAccount->setValue(QLatin1String("emailaddress"), account->fromAddress().address());
         ssoAccount->setValue(QLatin1String("fullName"), account->fromAddress().name());
